File: test_tracetools/src/test_intra.cpp
// Copyright 2019 Robert Bosch GmbH
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "test_tracetools/mark_process.hpp"

using namespace std::chrono_literals;

#define PUB_NODE_NAME "test_intra_pub"
#define SUB_NODE_NAME "test_intra_sub"
#define TOPIC_NAME "the_topic"

class PubIntraNode : public rclcpp::Node
{
public:
  explicit PubIntraNode(rclcpp::NodeOptions options)
  : Node(PUB_NODE_NAME, options)
  {
    pub_ = this->create_publisher<std_msgs::msg::String>(
      TOPIC_NAME,
      rclcpp::QoS(10));
    timer_ = this->create_wall_timer(
      500ms,
      std::bind(&PubIntraNode::timer_callback, this));
  }

private:
  void timer_callback()
  {
    auto msg = std::make_shared<std_msgs::msg::String>();
    msg->data = "some random intraprocess string";
    pub_->publish(*msg);
  }

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

class SubIntraNode : public rclcpp::Node
{
public:
  explicit SubIntraNode(rclcpp::NodeOptions options)
  : Node(SUB_NODE_NAME, options)
  {
    sub_ = this->create_subscription<std_msgs::msg::String>(
      "the_topic",
      rclcpp::QoS(10),
      std::bind(&SubIntraNode::callback, this, std::placeholders::_1));
  }

private:
  void callback(const std_msgs::msg::String::ConstSharedPtr msg)
  {
    RCLCPP_INFO(this->get_logger(), "[output] %s", msg->data.c_str());
    rclcpp::shutdown();
  }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
};

int main(int argc, char * argv[])
{
  test_tracetools::mark_trace_test_process();

  rclcpp::init(argc, argv);

  rclcpp::executors::SingleThreadedExecutor exec;
  auto pub_intra_node = std::make_shared<PubIntraNode>(
    rclcpp::NodeOptions().use_intra_process_comms(true));
  auto sub_intra_node = std::make_shared<SubIntraNode>(
    rclcpp::NodeOptions().use_intra_process_comms(true));
  exec.add_node(pub_intra_node);
  exec.add_node(sub_intra_node);

  printf("spinning\n");
  exec.spin();

  // Will actually be called inside the node's callback
  rclcpp::shutdown();
  return 0;
}

File: test_tracetools/test/test_intra.py
# Copyright 2019 Robert Bosch GmbH
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest

from tracetools_test.case import TraceTestCase
from tracetools_trace.tools import tracepoints as tp
from tracetools_trace.tools.lttng import is_lttng_installed


@unittest.skipIf(not is_lttng_installed(minimum_version='2.9.0'), 'LTTng is required')
class TestIntra(TraceTestCase):

    def __init__(self, *args) -> None:
        super().__init__(
            *args,
            session_name_prefix='session-test-intra',
            events_ros=[
                tp.rcl_subscription_init,
                tp.rclcpp_subscription_init,
                tp.rclcpp_subscription_callback_added,
                tp.callback_start,
                tp.callback_end,
            ],
            package='test_tracetools',
            nodes=['test_intra'],
        )

    def test_all(self):
        # Check events as set
        self.assertEventsSet(self._events_ros)

        # Check rcl_subscription_init events
        rcl_sub_init_events = self.get_events_with_name(tp.rcl_subscription_init)
        rcl_sub_init_topic_event = self.get_event_with_field_value_and_assert(
            'topic_name',
            '/the_topic',
            rcl_sub_init_events,
            allow_multiple=False,
        )

        # Get subscription handle
        sub_handle_intra = self.get_field(rcl_sub_init_topic_event, 'subscription_handle')

        # Check rclcpp_subscription_init events
        rclcpp_sub_init_events = self.get_events_with_name(tp.rclcpp_subscription_init)
        rclcpp_sub_init_topic_events = self.get_events_with_field_value(
            'subscription_handle',
            sub_handle_intra,
            rclcpp_sub_init_events,
        )
        # Should have 2 events for the given subscription handle
        # (Subscription and SubscriptionIntraProcess)
        self.assertNumEventsEqual(rclcpp_sub_init_topic_events, 2)

        # Get the 2 subscription pointers
        events = rclcpp_sub_init_topic_events
        subscription_pointers = [self.get_field(e, 'subscription') for e in events]

        # Get the corresponding callback pointers
        rclcpp_sub_callback_added_events = self.get_events_with_name(
            tp.rclcpp_subscription_callback_added,
        )
        rclcpp_sub_callback_added_topic_events = self.get_events_with_field_value(
            'subscription',
            subscription_pointers,
            rclcpp_sub_callback_added_events,
        )
        events = rclcpp_sub_callback_added_topic_events
        callback_pointers = [self.get_field(e, 'callback') for e in events]

        # Get corresponding callback start/end pairs
        start_events = self.get_events_with_name(tp.callback_start)
        end_events = self.get_events_with_name(tp.callback_end)
        # Should have at least one start:end pair
        self.assertNumEventsGreaterEqual(start_events, 1)
        self.assertNumEventsGreaterEqual(end_events, 1)
        start_events_topic = self.get_events_with_field_value(
            'callback',
            callback_pointers,
            start_events,
        )
        end_events_topic = self.get_events_with_field_value(
            'callback',
            callback_pointers,
            end_events,
        )
        self.assertNumEventsGreaterEqual(start_events_topic, 1)
        self.assertNumEventsGreaterEqual(end_events_topic, 1)

        # Check for is_intra_process field value of 1/true
        start_events_intra = self.get_events_with_field_value(
            'is_intra_process',
            1,
            start_events_topic,
        )
        # Should have one event
        self.assertNumEventsEqual(start_events_intra, 1)
        # As a sanity check, make sure its callback pointer has a corresponding callback_end event
        callback_pointer_intra = self.get_field(start_events_intra[0], 'callback')
        end_events_intra = self.get_events_with_field_value(
            'callback',
            callback_pointer_intra,
            end_events_topic,
        )
        self.assertNumEventsEqual(end_events_intra, 1)


if __name__ == '__main__':
    unittest.main()
